### PR TITLE
Optimize query 2

### PR DESCRIPTION
**Pull request by sargun**
It turns out testing on your laptop is never quite as good as testing in production.

[Please see commit messages for more information]

### CODE REVIEW

File: vpc/service/assign_addresses_v3.go [hunk 1]
```diff
@@ -1061,40 +1060,24 @@ func assignArbitraryIPv6AddressV3(ctx context.Context, tx *sql.Tx, branchENI *ec
...
 
-	if unusedIPAddresses.Len() > 0 {
+	if l := unusedIPAddresses.Len(); l > 0 {
 		unusedIPv6AddressesList := unusedIPAddresses.List()
 
-		rows, err := tx.QueryContext(ctx, "SELECT ip_address, last_seen FROM ip_last_used_v3 WHERE host(ip_address) = any($1) AND vpc_id = $2", pq.Array(unusedIPv6AddressesList), aws.StringValue(branchENI.VpcId))
-		if err != nil {
-			err = status.Error(codes.Unknown, errors.Wrap(err, "Could not fetch utilized IPv4 addresses from the database").Error())
-			span.SetStatus(traceStatusFromError(err))
+		row := tx.QueryRowContext(ctx, "SELECT ip_address FROM ip_last_used_v3 WHERE ip_address = any($1::inet[]) AND vpc_id = $2 ORDER BY last_seen ASC LIMIT 1", pq.Array(unusedIPv6AddressesList), aws.StringValue(branchENI.VpcId))
```
Comment (solarkennedy) [inline vpc/service/assign_addresses_v3.go:1066]: Just to confirm, this query picks the oldest ipv6 out of the database that has been last_seen, that happens to be already assigned to this agent, just not used by any task?

File: vpc/service/assign_addresses_v3.go [hunk 2]
```diff
@@ -1061,40 +1060,24 @@ func assignArbitraryIPv6AddressV3(ctx context.Context, tx *sql.Tx, branchENI *ec
...
+		var ipAddress string
+		err = row.Scan(&ipAddress)
+		if err == sql.ErrNoRows {
+			// Effectively choose a random one.
+			ipAddress = unusedIPv6AddressesList[rand.Intn(l)]
```
Comment (solarkennedy) [inline vpc/service/assign_addresses_v3.go:1071]: And if there are no rows, that means it is safe to pick one from the list we already have?

File: vpc/service/assign_addresses_v3.go [hunk 3]
```diff
@@ -1061,40 +1060,24 @@ func assignArbitraryIPv6AddressV3(ctx context.Context, tx *sql.Tx, branchENI *ec
...
+		} else if err != nil {
+			err = errors.Wrap(err, "Could not fetch utilized IPv6 addresses from the database")
+			tracehelpers.SetStatus(err, span)
 			return nil, err
 		}
```
Comment (solarkennedy) [inline vpc/service/assign_addresses_v3.go:1076]: And I guess there is the unspoken else, that there *is* an old ipv6 that was in our local unused list that was also previously seen, and that was already assigned to `ipAddress` during the scan.

**Review by solarkennedy — APPROVED** [review]
If my comments are correct in my understanding (I think I already get the `host()` optimization) then shipit!

Great job!

File: vpc/service/assign_addresses_v3.go (hunk 1 above)
Comment (sargun) [reply vpc/service/assign_addresses_v3.go:1066]: Correct.

File: vpc/service/assign_addresses_v3.go (hunk 2 above)
Comment (sargun) [reply vpc/service/assign_addresses_v3.go:1071]: Yes

Comment (sargun) [reply vpc/service/assign_addresses_v3.go:1071]: Well, one at random.

File: vpc/service/assign_addresses_v3.go (hunk 3 above)
Comment (sargun) [reply vpc/service/assign_addresses_v3.go:1076]: Well, if err == nil, then the scan was successful, and we successfully read the right IP.

**Pull request merged**